sklearn/tests/test_metadata_routing.py

```diff
@@ -1120,5 +1125,51 @@ def transform(self, X, sample_weight=None):
         def inverse_transform(self, X, sample_weight=None):
             pass
 
-    for method in METHODS:
+    # composite methods shouldn't have a corresponding set method.
+    for method in COMPOSITE_METHODS:
+        assert not hasattr(SimpleEstimator(), f"set_{method}_request")
+
+    # simple methods should have a corresponding set method.
+    for method in SIMPLE_METHODS:
         assert hasattr(SimpleEstimator(), f"set_{method}_request")
+
+
+def test_composite_methods():
+    class SimpleEstimator(BaseEstimator):
+        # This class should have every set_{method}_request
+        def fit(self, X, y, foo=None, bar=None):
+            pass
+
+        def predict(self, X, foo=None, bar=None):
+            pass
+
+        def transform(self, X, other_param=None):
+            pass
+
+    est = SimpleEstimator()
+    est.get_metadata_routing().fit_transform._requests == {}
+    est.get_metadata_routing().fit_predict._requests == {}
+
+    # setting the request on only one of them should raise an error
+    est.set_fit_request(foo=True, bar="test")
+    with pytest.raises(ValueError, match="Conflicting metadata requests for"):
+        est.get_metadata_routing().fit_predict
+
+    # setting the request on the other one should fail if not the same as the
+    # first method
+    est.set_predict_request(bar=True)
```

Comment:
I'm trying to see if there is a use case where this is too restrictive. What is the downside for allowing `fit` and `predict` to be in-consistent?

Reply:
ATM there's no overlap between the two methods, so the impact is kinda zero for existing scikit-learn code.

However, you can imagine `fit` and `transform` both accepting `sample_weight`, and wanting to pass different values to them. Then the code fails. But I kinda rather fix that if it happens, otherwise this part of the code would be quite a bit more complicated.